weave/trace_server/clickhouse_trace_server_batched.py

```diff
@@ -302,6 +307,42 @@ def calls_query_stream(
                 _ch_call_dict_to_call_schema_dict(ch_dict)
             )
 
+    def calls_delete(self, req: tsi.CallsDeleteReq) -> tsi.CallsDeleteRes:
+        if len(req.call_ids) > MAX_DELETE_CALLS_COUNT:
+            raise RequestTooLarge(
+                f"Cannot delete more than {MAX_DELETE_CALLS_COUNT} calls at once"
+            )
+
+        deleted_at = datetime.datetime.now()
+        proj_cond = "project_id = {project_id: String}"
+        proj_params = {"project_id": req.project_id}
+
+        # get all parents
+        parents = self._select_calls_query(
+            req.project_id,
+            conditions=[proj_cond, "id IN {ids: Array(String)}"],
+            parameters=proj_params | {"ids": req.call_ids},
+        )
+
+        # get all calls with trace_ids matching parents
+        all_calls = self._select_calls_query(
+            req.project_id,
+            conditions=[proj_cond, "trace_id IN {trace_ids: Array(String)}"],
+            parameters=proj_params | {"trace_ids": [p.trace_id for p in parents]},
+        )
+
+        all_descendants = find_call_descendants(
+            root_ids=req.call_ids,
+            all_calls=all_calls,
+        )
+
+        # create insert payload for all descendants (and parents)
+        data = [(req.project_id, d, deleted_at) for d in all_descendants]
+        self._insert(
```

Comment:
nit: while not necessary, it would be nice for this to use `_insert_call` instead of `_insert`. `_insert` feels risky to me because it is raw, untyped. `_insert_call` would require you to define a `CallDeleteCHInsertable` tp add to `CallCHInsertable = typing.Union[CallStartCHInsertable, CallEndCHInsertable]`

The reason i like this pattern more is that it enforces the expected schema of a delete. This is important because we absolutely must be sure that the insert only contains the fields `"project_id", "id", "deleted_at"`... which is enforced by this list right now. Why? Well the aggregation functions such as `anySimpleState(op_name)` will take any-non-null value... meaning the delete entries could inadvertently mess with other fields. This is more of the theoretical callout, so consider it optional.